accounts/accounts/controllers/registration.py

```diff
@@ -0,0 +1,303 @@
+    """Handle requests to view a user's profile."""
+    user = users.get_user_by_id(user_id)
+    return {'user': user}, status.HTTP_200_OK, {}
+
+
+def edit_profile(method: str, user_id: str, session: domain.Session,
+                 params: Optional[MultiDict] = None,
+                 ip: Optional[str] = None) -> ResponseData:
+    """Handle requests to update a user's profile."""
+    if method == 'GET':
+        user = users.get_user_by_id(user_id)
+        form = ProfileForm.from_domain(user)
+        data = {'form': form, 'user_id': user_id}
+    elif method == 'POST':
+        form = ProfileForm(params)
+        data = {'form': form, 'user_id': user_id}
+
+        if not form.validate():
+            return data, status.HTTP_400_BAD_REQUEST, {}
+
+        if form.user_id.data != user_id:
+            msg = 'User ID in request does not match'
+            raise BadRequest(msg)  # type: ignore
+
+        user = form.to_domain()
+        try:
+            user, auth = users.update(user)
+        except Exception as e:
+            data['error'] = 'Could not save user profile; please try again'
+            return data, status.HTTP_500_INTERNAL_SERVER_ERROR, {}
+
+        # We need a new session, to update user's data.
+        _logout(session.session_id)
+        new_session, new_cookie = _login(user, auth, ip)
+        data.update({'cookies': {
+            'session_cookie': (new_cookie, new_session.expires)
+        }})
+        return data, status.HTTP_303_SEE_OTHER, {}
+    return data, status.HTTP_200_OK, {}
+
+
+class ProfileForm(Form):
+    """User registration form."""
+
+    COUNTRIES = [('', '')] + \
+        [(country.alpha_2, country.name) for country in pycountry.countries]
+    RANKS = [('', '')] + domain.UserProfile.RANKS
+    GROUPS = [
+        (key, group['name']) for key, group in taxonomy.GROUPS.items()
+        if not group.get('is_test', False)
+    ]
+    CATEGORIES = [
+        (archive['name'], [
+            (category_id, category['name'])
+            for category_id, category in taxonomy.CATEGORIES_ACTIVE.items()
+            if category['in_archive'] == archive_id
+        ])
+        for archive_id, archive in taxonomy.ARCHIVES_ACTIVE.items()
+    ]
+    """Categories grouped by archive."""
+
+    user_id = HiddenField('User ID')
+
+    email = StringField(
+        'Email address',
+        validators=[Email(), Length(max=255), DataRequired()],
+        description="You must be able to receive mail at this address."
+        " We take <a href='https://arxiv.org/help/email-protection'>"
+        " strong measures</a> to protect your email address from viruses and"
+        " spam. Do not enter an e-mail address that belongs to someone"
+        " else: if we discover that you've done so, we will suspend your"
+        " account."
+    )
+
+    username = StringField('Username',
+                           validators=[Length(min=5, max=20), DataRequired()])
+    forename = StringField('First or given name',
```

Comment:
The classic system will push back (albeit gently) if it finds an existing account with the same forename+surname. 
<img width="1022" alt="screenshot 2018-07-09 10 39 08" src="https://user-images.githubusercontent.com/746253/42457096-83d24ec0-8364-11e8-9794-0e797af7e068.png">

Reply:
Great catch; created https://culibrary.atlassian.net/browse/ARXIVNG-973